kusto/cloudinfo_test.go

```diff
@@ -0,0 +1,93 @@
+				loginEndpoint:          "https://login.microsoftonline.com",
+				loginMfaRequired:       false,
+				kustoClientAppId:       "db662dc1-0cfe-4e1c-a843-19a68e65be58",
+				kustoClientRedirectUri: "https://microsoft/kustoclient",
+				kustoServiceResourceId: "https://kusto.dev.kusto.windows.net",
+				firstPartyAuthorityUrl: "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a",
+			},
+		},
+		{
+			name:            "test_login_endpoint_2",
+			responsePayload: `{"AzureAD": {"LoginEndpoint": "https://login2.microsoftonline.com","LoginMfaRequired": true,"KustoClientAppId": "db662dc1-0cfe-4e1c-a843-19a68e65bxxx","KustoClientRedirectUri": "https://microsoft/kustoclient","KustoServiceResourceId": "https://kusto.dev.kusto.windows.net","FirstPartyAuthorityUrl": "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e912xxx"  },  "dSTS": {"CloudEndpointSuffix": "windows.net","DstsRealm": "realm://dsts.core.windows.net","DstsInstance": "prod-dsts.dsts.core.windows.net","KustoDnsHostName": "kusto.windows.net","ServiceName": "kusto"}}`,
+			expectedCloudInfo: &cloudInfo{
+				loginEndpoint:          "https://login2.microsoftonline.com",
+				loginMfaRequired:       true,
+				kustoClientAppId:       "db662dc1-0cfe-4e1c-a843-19a68e65bxxx",
+				kustoClientRedirectUri: "https://microsoft/kustoclient",
+				kustoServiceResourceId: "https://kusto.dev.kusto.windows.net",
+				firstPartyAuthorityUrl: "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e912xxx",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			server := runServer(http.StatusOK, tt.responsePayload, true)
+			actualCloudInfo, err := RetrieveCloudInfoMetadata(server.URL)
+			assert.Nil(t, err)
+			assert.Equal(t, tt.expectedCloudInfo, actualCloudInfo)
+			server.Close()
+		})
+	}
+}
+
+func TestRetrieveCloudInfoMetadataError(t *testing.T) {
+	responsePayload := ``
+	server := runServer(http.StatusInternalServerError, responsePayload, false)
+	actualCloudInfo, err := RetrieveCloudInfoMetadata(server.URL)
+	server.Close()
+	assert.NotNil(t, err)
+	assert.Nil(t, actualCloudInfo)
+	errorMessage := err.Error()
+	assert.EqualValues(t, fmt.Sprintf("retrieved error code %d when querying endpoint %s/%s", http.StatusInternalServerError, server.URL, metadataEndpoint), errorMessage)
+}
+
+func TestRetrieveCloudInfoMetadataNotFound(t *testing.T) {
+	server := runServer(http.StatusNotFound, ``, false)
+	actualCloudInfo, err := RetrieveCloudInfoMetadata(server.URL)
+	server.Close()
+	assert.Nil(t, err)
+	assert.Equal(t, defaultCloudInfo, actualCloudInfo)
+}
+
+func TestRetrieveCloudInfoMetadataSuccessNoBody(t *testing.T) {
+	server := runServer(http.StatusOK, ``, true)
+	actualCloudInfo, err := RetrieveCloudInfoMetadata(server.URL)
+	server.Close()
+	assert.Nil(t, err)
+	assert.Equal(t, defaultCloudInfo, actualCloudInfo)
+}
+
+func runServer(statusCode int, payload string, isSuccess bool) *httptest.Server {
```

Comment:
To combine the tests above into a single table driven test, you can expand this.

```go
type server struct {
	code    int
	payload []byte
	http    *httptest.Server
}

func newTestServ() *server {
	s := &server{}
	s.http = httptest.NewServer(s)
	return s
}

func (s *server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	log.Println("got here")
	defer log.Println("exited")
	w.WriteHeader(s.code)
	if s.code != 200 {
		w.Write(s.payload)
	}
}

func (s *server) urlStr() string {
	return s.http.URL
}

func (s *server) close() {
	s.http.Close()
}
```

Then you can use it in each iteration of the test by:

```go
         s := newTestServ()
	defer s.close()

        for _, test := range tests {
	  s.code = test.code
	  s.payload = test.payload
          ...

          resp, err := Get(s.urlStr())
          switch {
          case err != nil && !test.err:
            t.Errorf("TestGet(%s): got err == %s, want err == nil", test.desc, err)
            continue
          case err == nil && test.err:
            t.Errorf("TestGet(%s): got err == nil, want err != nil", test.desc)
            continue
          case err != nil:
            continue
          }
         // Do assert stuff here
        }
        }
```